pkg/ingester/ingester_v2.go

```diff
@@ -1280,3 +1312,435 @@ func (i *Ingester) v2FlushHandler(w http.ResponseWriter, _ *http.Request) {
+			}
+
+			// Limit the number of TSDB's opening concurrently. Start blocks until there's a free spot available or the context is cancelled.
+			if err := openGate.Start(ctx); err != nil {
+				runErr = err
+				break
+			}
+
+			wg.Add(1)
+			go func(userID, bucketName string) {
+				defer wg.Done()
+				defer openGate.Done()
+				defer func(ts time.Time) {
+					i.TSDBState.walReplayTime.Observe(time.Since(ts).Seconds())
+				}(time.Now())
+
+				start, end, err := getBucketRangesForBucketName(bucketName)
+				db, err := i.createBackfillTSDB(userID, bucketName, int64(end-start)*2)
+				if err != nil {
+					level.Error(util.Logger).Log("msg", "unable to open user backfill TSDB", "err", err, "user", userID)
+					return
+				}
+
+				bucket := &tsdbBucket{
+					db:          db,
+					lastAppend:  time.Now(),
+					bucketStart: start,
+					bucketEnd:   end,
+				}
+
+				i.TSDBState.backfillDBs.userStateMtx.Lock()
+				// Append at the end, we will sort it at the end.
+				i.TSDBState.backfillDBs.tsdbs[userID] = append(i.TSDBState.backfillDBs.tsdbs[userID], bucket)
+				i.TSDBState.backfillDBs.userStateMtx.Unlock()
+			}(userID, bucketName.Name())
+		}
+
+		if runErr != nil {
+			break
+		}
+
+	}
+
+	// Wait for all opening routines to finish
+	wg.Wait()
+
+	// Sort the buckets within the users.
+	i.TSDBState.backfillDBs.userStateMtx.Lock()
+	for _, buckets := range i.TSDBState.backfillDBs.tsdbs {
+		sort.Slice(buckets, func(i, j int) bool {
+			return buckets[i].bucketStart < buckets[i].bucketEnd
+		})
+	}
+	i.TSDBState.backfillDBs.userStateMtx.Unlock()
+
+	return runErr
+}
+
+func (i *Ingester) backfillSelect(ctx context.Context, userID string, from, through int64, matchers []*labels.Matcher) ([]storage.SeriesSet, error) {
+	buckets := i.TSDBState.backfillDBs.getBucketsForUser(userID)
+
+	var queriers []storage.Querier
+	defer func() {
+		for _, q := range queriers {
+			q.Close()
+		}
+	}()
+	for _, b := range buckets {
+		if !b.overlaps(from, through) {
+			mint, err := b.db.DB.StartTime()
+			if err != nil {
+				return nil, err
+			}
+			maxt := b.db.Head().MaxTime()
+			if !overlapsOpenInterval(mint, maxt, from, through) {
+				continue
+			}
+		}
+
+		q, err := b.db.Querier(ctx, from, through)
+		if err != nil {
+			return nil, err
+		}
+
+		queriers = append(queriers, q)
+	}
+
+	if len(queriers) == 0 {
+		return nil, nil
+	}
+
+	result := make([]storage.SeriesSet, len(queriers))
+	errC := make(chan error, 1)
+	var wg sync.WaitGroup
+	for i, q := range queriers {
+		wg.Add(1)
+		go func(i int, q storage.Querier) {
+			defer wg.Done()
+
+			ss := q.Select(false, nil, matchers...)
+			if ss.Err() != nil {
+				select {
+				case errC <- ss.Err():
+				default:
+				}
+			}
+			result[i] = ss
+		}(i, q)
+	}
+
+	wg.Wait()
+	select {
+	case err := <-errC:
+		return nil, err
+	}
+
+	return result, nil
+}
+
+type tsdbBucket struct {
+	db                     *userTSDB
+	bucketStart, bucketEnd model.Time
+	lastAppend             time.Time
```

Comment:
You already have this in `userTSDB`. We could remove this and use `userTSDB.lastUpdate`